src/rest.ts

```diff
@@ -55,39 +61,55 @@ function splitAtNLNL(array) {
     return [array, null]; // not-covered: ditto
 }
 
-/* uid: null for logged in session user; 0 for root; in the future we'll support other users
- * Returns a connection object with methods monitor(), call(), and close(), and an `uid` property.
- */
-function connect(uid) {
+export type MonitorCallbackJson = (data: JsonObject) => void;
+export type MonitorCallbackRaw = (data: Uint8Array) => void;
+export type MonitorCallback = MonitorCallbackJson | MonitorCallbackRaw;
+
+// type predicate helper for narrowing which monitor callback is being used
+function isReturnRaw(return_raw: boolean, callback: MonitorCallback): callback is MonitorCallbackRaw {
+    return return_raw;
+}
+
+export type Connection = {
+    uid: Uid;
+    monitor: (path: string, callback: MonitorCallback, return_raw?: boolean) => Promise<void>;
+    call: (options: JsonObject) => Promise<string>;
+    close: () => void;
+};
+
+function connect(uid: Uid): Connection {
     const addr = getAddress(uid);
     /* This doesn't create a channel until a request */
     /* HACK: use binary channel to work around https://github.com/cockpit-project/cockpit/issues/19235 */
     const http = cockpit.http(addr.path, { superuser: addr.superuser, binary: true });
-    const raw_channels = [];
-    const connection = { uid };
+    const raw_channels: cockpit.Channel<Uint8Array>[] = [];
+    const encoder = new TextEncoder();
     const decoder = new TextDecoder();
     const user_str = (uid === null) ? "user" : (uid === 0) ? "root" : `uid ${uid}`;
 
-    connection.call = function (options) {
+    function call(options: JsonObject): Promise<string> {
         const id = call_id++;
         debug(user_str, `call ${id}:`, JSON.stringify(options));
         return new Promise((resolve, reject) => {
             options = options || {};
             http.request(options)
-                    .then(result => {
+                    .then((result: Uint8Array) => {
                         const text = decoder.decode(result);
                         debug(user_str, `call ${id} result:`, text);
                         resolve(text);
                     })
-                    .catch((error, content) => {
-                        const text = decoder.decode(content);
-                        debug(user_str, `call ${id} error:`, JSON.stringify(error), "content", text);
-                        manage_error(reject, error, text);
+                    // @ts-expect-error: magic cockpit defer error extra "content" parameter
+                    .catch((error: object, content: unknown) => {
+                        const content_text = (content instanceof Uint8Array)
+                            ? decoder.decode(content as Uint8Array)
+                            : content;
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2214-c7aa3398-20250801-083853-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)

Reply:
podman API contract.